### DISCUSSION TITLE
Recommended patterns to allow non-threadsafe primitives in the host env context?

### DISCUSSION THREAD

**Discussion by kwonoj** [wasmerio/wasmer#2830] [Q&A]
Roughly, this is what I'm trying to achieve:

```
struct SomeSingleThreadOnly {
 leading: Rc<RefCell<HashMap<BytePos, Vec<SomeStruct>, BuildHasherDefault<FxHasher>>>>
...
}

fn init<'a>(
value: Option<&'a SomeSingleThreadOnly>
) {
let wasmer_store = module.store();

let fn_decl = Function::new_native_with_env(
                wasmer_store,
                HostEnvironment {
                    memory: LazyInit::default(),
                    value: // uggggggh...
                },
                fn_decl_impl
            );

}
```

Unfortunately passed param (`Option<&'a SomeSingleThreadOnly>`) is something not easily controllable / or changable, and given `WasmerEnv` doesn't allow to set non-thread safety values to its environment there's no easy way to make compiler satisfied.

Logically in our execution path we _know_ all of wasm execution is sequential, and there's no paths can create race conditions, while it is not possible to tell wasmerEnv to treat so. Is there any recommended pattern to force escape constraints?

**Comment by Amanieu** [comment]
Currently this is not possible with Wasmer's API, you need to use `Arc<Mutex>` instead of `Rc<RefCell>`.

We are looking into ways to improve this in future versions of Wasmer.
